webrender/src/render_backend.rs

```diff
@@ -310,7 +318,21 @@ impl RenderBackend {
         let mut new_pipeline_sizes = HashMap::new();
 
         if let Some(id) = self.current_bound_webgl_context_id {
-            self.webgl_contexts.get(&id).unwrap().unbind().unwrap();
+            self.webgl_contexts.get(&id).unwrap().unbind();
+            self.current_bound_webgl_context_id = None;
+        }
+
+        // When running in OSMesa mode with texture sharing,
+        // a flush is required on any GL contexts to ensure
+        // that read-back from the shared texture returns
+        // valid data! This should be fine to have run on all
+        // implementations - a single flush for each webgl
+        // context at the start of a render frame should
+        // incur minimal cost.
+        for (_, webgl_context) in &self.webgl_contexts {
+            webgl_context.make_current();
+            gl::flush();
```

Comment:
Sure, sounds fine, r=me with a comment about being able to do it less conservatively :)